### DISCUSSION TITLE
Difference between UaSubscription.NotificationListener and UaSubscriptionManager.SubscriptionListener

### DISCUSSION THREAD

**Discussion by dcntrd** [eclipse-milo/milo#1224] [General]
What is the difference between the two listeners. At first glance they have the same methods. The only difference I see is that the UaSubscription.NotificationListener can be added to the Subscription and the UaSubscriptionManager.SubscriptionListener can be added to the SubscriptionManager - but is there any other relevant difference between the two listeners?

**Comment by dcntrd** [comment]
... and when is the onStatusChangedNotification() called?

I've added a subscription and a NotificationListener, but I could not observer any call to the onStatusChangedNotification(). I've expected at least one call when the subscription is initialized and another one when e.g. shutting down the server

**Reply by kevinherron** [reply]
There's a bit of overlap. One listener will receive callbacks for all subscriptions, one for just the subscription it was added to.

StatusChangeNotifications are a specific type of notification sent by the server and AFAIK the spec only uses them in 2 scenarios:
1. the Subscription has timed out (Bad_Timeout)
2. the Subscription was transferred (Good_Transferred)

It's not as useful as it sounds in a general purpose sense.

**Reply by dcntrd** [reply]
Thank you for clarification

**Reply by kevinherron** [reply]
FWIW these confusing Subscription and MonitoredItem APIs are being reworked for 1.0.

You can see how it's shaping up here if you're interested: https://github.com/eclipse/milo/pull/1223

https://github.com/eclipse/milo/blob/00e3090342593be3a169ef04714bbe64c8b03b04/milo-examples/client-examples/src/main/java/org/eclipse/milo/examples/client/SubscriptionDataExample.java#L37-L93
